TEZ-3998: support constructing DAG with concurrent edge

yingdachen opened the pull request:
1. unblock the construction of DAG with concurrent edge
2. introduce VerteManagerWithConcurrentInput with SilentEdgeManager
3. introcuce the concept of ConcurrentEdgeTriggerType
4. UT
Change-Id: I4202e2946b99c440d7d7923151588632d74121cc

yingdachen commented:
@jteagles do you have any recommendation about who should this pull request be assigned to?

Thanks,

yingdachen commented:
> @yingdachen since this is part of a larger end-to-end feature (Tez-3997) would it be better for you to own a feature branch, in which you commit for the subtasks, and later, when the feature is complete, merge into master?
> It is easier to provide meaningful feedback when seeing how it all comes together - right now, most of the changes are stubs.

@anicoara thanks for the feedback.

couple of things

1. The design doc in 3997 was meant to provide the big picture and as the place to collect feedback for the overall design. We believe suitable level of details about proposed changes have been provided, and it would be great hear from community on the design.

2. The feature was broken into 4 tasks carefully that range from basic support and plugin addition(3998), api change in AM component (3999), change in runtime (4000) and more complete scenario support (4001). We believe such break-down is suitable for us as new contributors to fit in, beginning with a simple change discussed here (3998) that is standalone in itself. We are hoping that our changes will take a gradual path to completion, which would not necessitate a separate feature branch.

3. In the change introduced here, the SilentEdgeManager is indeed a stub, that is by design and would not likely to change. However, I would not categorize the other change (such as the new VertexManagerWithConcurrentInput) as so, since it is functional as it is and we have added UT coverage to cover that as well.

jteagles commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,234 @@
+
+  @Override
+  public void onVertexStateUpdated(VertexStateUpdate stateUpdate) {
+    Preconditions.checkArgument(VertexState.CONFIGURED.equals(stateUpdate.getVertexState()),
```
Precondition performance may be a consideration here since message is string is composed regardless of boolean condition. Perhaps replace with 3 argument Precondition or evaluate condition and throw IllegalArgumentExcteption manually, constructing the message in the condition block.
Same for other precondition checks.
https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Preconditions.html#checkArgument(boolean,%20java.lang.String,%20java.lang.Object...)

yingdachen replied:
done, replace with check-first-and-throw-if-necessary

jteagles commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,234 @@
+
+  @Override
+  public void onSourceTaskCompleted(TaskAttemptIdentifier attempt) {
+    completedUpstreamTasks ++;
```
```suggestion
    completedUpstreamTasks++;
```

yingdachen replied:
done

beltran replied:
Maybe log the value of this variable in the log below, something like : completed/total.

jteagles commented on tez-runtime-library/src/test/java/org/apache/tez/dag/library/vertexmanager/TestVertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,114 @@
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @Test//(timeout=5000)
```
Perhaps timeout should be uncommented
```suggestion
  @Test(timeout=5000)
```

yingdachen replied:
done

jteagles commented:
@yingdachen, a few other considerations. 1)Normally, a feature like this could be a candidate for feature branch. Do to the low risk interaction with Systems and backwards compatible API, I think we could allow develop in trunk. It may cause some coordinating with ongoing 0.10 build to prevent delay of that release line. 2) Please provide some input on fault tolerance of AM and attempt in the presence of Concurrent Edges. Specifically, we will want to provide fault tolerance guarantees and possible interactions with AM recovery and assign component responsibilities and interactions where needed.

yingdachen commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,234 @@
+    managedTasks = getContext().getVertexNumTasks(getContext().getVertexName());
+    Map<String, EdgeProperty> edges = getContext().getInputVertexEdgeProperties();
+    for (Map.Entry<String, EdgeProperty> entry : edges.entrySet()) {
+      if (!CONCURRENT.equals(entry.getValue().getSchedulingType())){
```
done

yingdachen commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,234 @@
+
+import static org.apache.tez.dag.api.EdgeProperty.SchedulingType.CONCURRENT;
+
+public class VertexManagerWithConcurrentInput extends VertexManagerPlugin{
```
done

yingdachen commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/edgemanager/SilentEdgeManager.java:
```diff
@@ -0,0 +1,89 @@
+ * including both routing and actual data transfers.
+ */
+
+public class SilentEdgeManager extends EdgeManagerPlugin{
```
done

yingdachen commented:
for the latest patch

1. address previous code review comments
2. add check in DAGAppMaster, for a DAG with concurrent edge, disable recovery (enforce the recovered DAG to run from scratch in case of AM failover). We will tackle proper failover strategy for a DAG with concurrent edge in TEZ-4017.

beltran commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,242 @@
+
+  @Override
+  public void initialize() {
+    if (getContext().getUserPayload() == null) {
```
Should the same check as in [here](https://github.com/apache/tez/blob/master/tez-dag/src/main/java/org/apache/tez/dag/app/dag/impl/RootInputVertexManager.java#L235) be done? Otherwise a NullPointerException may happen at `vertexConfig = TezUtils.createConfFromUserPayload(getContext().getUserPayload());`.

beltran commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,242 @@
+  private Configuration vertexConfig;
+  private String vertexName;
+  private ConcurrentEdgeTriggerType edgeTriggerType;
+  private boolean allSrcVerticesConfigured;
```
This variable seems to be written/read in several callback. Maybe it's worth to make it volatile and prevent some race condition?

beltran commented on tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java:
```diff
@@ -0,0 +1,242 @@
+        break;
+      }
+    }
+    allSrcVerticesConfigured = checkAllSrcVerticesConfigured;
```
Doing it like this makes me wonder if the tasks can be scheduled twice for example `onVertexStateUpdated` being called twice closely:

* Thread 1: calls onVertexStateUpdated
* Thread 2: calls onVertexStateUpdated
* Thread 1: at `allSrcVerticesConfigured = checkAllSrcVerticesConfigured;` is set to false
* Thread 2: at `allSrcVerticesConfigured = checkAllSrcVerticesConfigured;` is set to true now
* Thread 1: scheduleTasks schedules all the tasks because  `allSrcVerticesConfigured` is true
* Thread 2: scheduleTasks schedules all the tasks because  `allSrcVerticesConfigured` is true

I see the variable `tasksScheduled` is used, but I think it could still happen. Maybe with some `AtomicBoolean.getAndSet` we'd be safer.

beltran replied:
Sorry I meant `compareAndSet` to make sure `getContext().scheduleTasks(tasksToStart)` is not called twice. I think now it could still happen. So it would be something like
```java
if (tasksScheduled .compareAndSet(false, true)) {
     // schedule tasks
}
```

yingdachen replied:
i see what you meant. done

beltran replied:
What I'm saying is that two threads can still go pass `if (tasksScheduled.get()) {` when it's `false`, then `if (!canScheduleTasks()) {` would be true for both threads and it doesn't matter if you `tasksScheduled.compareAndSet` if you don't check the return value.

yingdachen replied:
I think the source of risk here is from the plugin interfaces are not synchronized. I have made all meaningful implementation of the plugin interfaces (onVertexStateUpdated, onSourceTaskCompleted, etc.) synchronized, following the common practice of other plugin implmentation (such as ShuffleVertexManagerBase).

We should be good now.

beltran replied:
I don't think it's necessary to synchronize all the methods and that would unnecessarily slow this down when an `AtomicBoolean` to guard `getContext().scheduleTasks(tasksToStart)` would be enough IMO. 
This does solve the bug though so I'd be happy to create a follow up ticket to address this and let this go in how it is in this PR. Maybe we can look into this again when we see the whole picture when the other tickets implemented.

beltran commented on tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java:
```diff
@@ -2440,23 +2453,30 @@ public void run() {
           TezConstants.TEZ_PB_PLAN_BINARY_NAME));
       dagPlan = DAGPlan.parseFrom(dagPBBinaryStream);
-
-      startDAG(dagPlan, null);
```
I see this has been moved [here](https://github.com/apache/tez/pull/33/files#diff-755c0ec043a1800cd6cbf31823a59c8fR2069). Are there going to be any consequences with that? The logic is not exactly the same since before `startDAG(dagPlan, null);` is called as long as `(!isSession)` and now it wouldn't if it gets in the section `if (recoveredDAGData != null) {` in `DAGAppMaster.serviceStart`

yingdachen replied:
I am not sure whether you are referring to the recoveredDAGData != null on line 1993 or 1983.

For the one on 1993, it would not call startDAG anyway. For the one on 1983, the logic added is only to forcefully invalid the recovery data that has been read, it does not impact the startDAG behavior either.

Overall, there are two modification here:
1.  the DAGPlan is read a bit earlier (in non-session mode)
2.  the already-read recovery content may be useless when there is concurrent edge in the DAG. This is done on purpose and I think should be acceptable since it does not modify existing logic flow and provide the safeguard for DAG execution with concurrent edge, before we have a proper failover implementation for that TEZ-4017

beltran replied:
Oh yes, just saw there are two lines like that. I was referring to the one in 1993. So before when `startDAG` was called and then got into 1993 it had no effect?

yingdachen replied:
1993 is the line for if (recoveredDAGData != null) , when inside that branch startDAG would not be called, it would only be called in the corresponding "else" branch

beltran replied:
OK I see, got confused with what was happening before.

yingdachen commented on tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java:
```diff
@@ -408,6 +409,18 @@ public TezConfiguration(boolean loadDefaults) {
 
+  /**
+   * String value. Describes the timing of scheduling downstream vertex tasks
+   * when the vertex has a concurrent input edge.
```
I have expand the comment a bit in an effort to make it easier to understand. 

Not using the event name directly is due to the subtle difference between event and scheduling actually being kicked off: for example, in the case of "SOURCE_TASK_STARTED", the scheduling is not triggered by any TASK_STARTED event, but instead it will only be triggered after enough TASK_STARTED events have been collected (for example, all the upstream tasks are running). For the lack of a better choice, I think defining a TriggerType is easier to expand in the long run too.

yingdachen commented:
@beltran Thanks for your detailed review, I have addressed your review comments in the new iteration.

@beltran @jteagles can you take another look at your convenience. 

thanks,

yingdachen commented:
@jteagles @beltran latest patch has addressed all previous code review comments, please take a look at your convenience. thanks

yingdachen commented:
@jteagles have you got any chance to proceed? thanks,
